apps/heft/src/plugins/FileGlobSpecifier.ts

```diff
@@ -129,39 +136,55 @@ export async function watchGlobAsync(
   return results;
 }
 
-export async function getFilePathsAsync(
-  fileGlobSpecifier: IFileSelectionSpecifier,
-  fs?: FileSystemAdapter
-): Promise<Set<string>> {
-  const rawFiles: string[] = await glob(fileGlobSpecifier.includeGlobs!, {
-    fs,
+export async function getFileSelectionSpecifierPathsAsync(
+  options: IGetFileSelectionSpecifierPathsOptions
+): Promise<Map<string, fs.Dirent>> {
+  const { fileGlobSpecifier, includeFolders, fileSystemAdapter } = options;
+  const rawEntries: Entry[] = await glob(fileGlobSpecifier.includeGlobs!, {
+    fs: fileSystemAdapter,
     cwd: fileGlobSpecifier.sourcePath,
     ignore: fileGlobSpecifier.excludeGlobs,
+    onlyFiles: !includeFolders,
     dot: true,
-    absolute: true
+    absolute: true,
+    objectMode: true
   });
 
-  if (fs && isWatchFileSystemAdapter(fs)) {
-    const changedFiles: Set<string> = new Set();
+  let results: Map<string, fs.Dirent>;
+  if (fileSystemAdapter && isWatchFileSystemAdapter(fileSystemAdapter)) {
+    results = new Map();
     await Async.forEachAsync(
-      rawFiles,
-      async (file: string) => {
-        const state: IWatchedFileState = await fs.getStateAndTrackAsync(path.normalize(file));
+      rawEntries,
+      async (entry: Entry) => {
+        const { path: filePath, dirent } = entry;
+        if (entry.dirent.isDirectory()) {
+          return;
+        }
+        const state: IWatchedFileState = await fileSystemAdapter.getStateAndTrackAsync(
+          path.normalize(filePath)
+        );
         if (state.changed) {
-          changedFiles.add(file);
+          results.set(filePath, dirent);
         }
       },
       {
         concurrency: 20
       }
     );
-    return changedFiles;
+  } else {
+    results = new Map(rawEntries.map((entry) => [entry.path, entry.dirent]));
   }
 
-  return new Set(rawFiles);
+  return results;
 }
 
-export function normalizeFileSelectionSpecifier(fileGlobSpecifier: IFileSelectionSpecifier): void {
+export function normalizeFileSelectionSpecifier(
+  rootPath: string,
+  fileGlobSpecifier: IFileSelectionSpecifier
+): void {
+  fileGlobSpecifier.sourcePath = fileGlobSpecifier.sourcePath
+    ? path.resolve(rootPath, fileGlobSpecifier.sourcePath)
+    : rootPath;
```

Comment:
```suggestion
  const { sourcePath } = fileGlobSpecifier;
  fileGlobSpecifier.sourcePath = sourcePath
    ? path.resolve(rootPath, sourcePath)
    : rootPath;
```
I find it makes it slightly more readable and clearer that the value is the same in both the condition and the resolve call.